crypto3/libs/blueprint/include/nil/blueprint/bbf/generic.hpp

```diff
@@ -172,44 +176,96 @@ namespace nil {
                 }
             };
 
+            // A class for storing the information on which cells in the assignment table is already allocated/used.
+            template<typename FieldType>
+            class allocation_log {
+            public:
+                using assignment_description_type = nil::crypto3::zk::snark::plonk_table_description<FieldType>;
+
+                allocation_log(const assignment_description_type& desc) {
+                    log[column_type::witness] = std::vector<std::vector<bool>>(
+                        desc.witness_columns, std::vector<bool>(desc.usable_rows_amount));
+                    log[column_type::public_input] = std::vector<std::vector<bool>>(
+                        desc.public_input_columns, std::vector<bool>(desc.usable_rows_amount));
+                    log[column_type::constant] = std::vector<std::vector<bool>>(
+                        desc.constant_columns, std::vector<bool>(desc.usable_rows_amount));
+                }
+
+                bool is_allocated(std::size_t col, std::size_t row, column_type t) {
+                    if (col >= log[t].size() || row >= log[t][col].size()) {
+                        std::stringstream error;
+                        error << "Invalid value col = " << col << ", row = " << row 
+                            << " when checking if a " << t << " cell is allocated. We have "
+                            << log[t].size() << " columns each having "
+                            << log[t][col].size() << " rows.";
```

Comment:
This will go out of range if we had col >= log[t].size(), so we should rather do smth like log[t][0].size(), but in fact we are not guaranteed to have columns of a specific type at all, so maybe we should just shorten this error report.

Reply:
Changed.